bpf/bpf_lxc.c

```diff
@@ -54,6 +54,69 @@ static __always_inline bool redirect_to_proxy(int verdict, __u8 dir)
 #endif
 
 #ifdef ENABLE_IPV6
+# ifndef ENABLE_HOST_SERVICES_FULL
+declare_tailcall_if(__and(is_defined(ENABLE_IPV4), is_defined(ENABLE_IPV6)),
+		    CILIUM_CALL_IPV6_LOCAL_FROM_LXC)
+int tail_handle_ipv6_local(struct __ctx_buff *ctx)
+{
+	struct lb6_service *svc;
+	struct lb6_key key = {};
+	struct csum_offset csum_off = {};
+	struct ct_state ct_state_new = {};
+	struct ipv6_ct_tuple tuple = {};
+	void *data, *data_end;
+	struct ipv6hdr *ip6;
+	int ret, l4_off, hdrlen;
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip6))
+		return DROP_INVALID;
+
+	ipv6_addr_copy(&tuple.daddr, (union v6addr *)&ip6->daddr);
+	ipv6_addr_copy(&tuple.saddr, (union v6addr *)&ip6->saddr);
+	tuple.nexthdr = ip6->nexthdr;
+
+	hdrlen = ipv6_hdrlen(ctx, ETH_HLEN, &tuple.nexthdr);
+	if (hdrlen < 0)
+		return hdrlen;
+
+	l4_off = ETH_HLEN + hdrlen;
+
+	ret = lb6_extract_key(ctx, &tuple, l4_off, &key, &csum_off,
+			      CT_EGRESS);
+	if (IS_ERR(ret)) {
+		if (ret == DROP_UNKNOWN_L4)
+			goto recirculate_tailcall;
+		goto drop_err;
+	}
+
+	/*
+	 * Check if the destination address is among the address that should
+	 * be load balanced. This operation is performed before we go through
+	 * the connection tracker to allow storing the reverse nat index in
+	 * the CT entry for destination endpoints where we can't encode the
+	 * state in the address.
+	 */
+	svc = lb6_lookup_service(&key, is_defined(ENABLE_NODEPORT));
+	if (svc) {
+		ret = lb6_local(get_ct_map6(&tuple), ctx, ETH_HLEN, l4_off,
+				&csum_off, &key, &tuple, svc, &ct_state_new);
+		if (IS_ERR(ret))
+			goto drop_err;
+	}
+
+recirculate_tailcall:
+	bpf_skip_lb6_local_set(ctx);
+	ctx_store_meta_ct_state6(ctx, &ct_state_new);
+
+	ep_tail_call(ctx, CILIUM_CALL_IPV6_FROM_LXC);
+	ret = DROP_MISSED_TAIL_CALL;
```

Comment:
Depending on the conditions passed, `declare_tailcall_if` might be a function declaration (no tail call), in which case we don't want the above code to be executed.